### ISSUE TITLE
Have you tried without softmax?

### ISSUE THREAD

**Issue by yuzhobotianzhiyu** [thuml/MobileAttention#1]
Because at the deployment level, softmax is very time-consuming. 
”source_competition = torch.softmax(conserved_source, dim=-1) * float(key.shape[2])“

I found that softmax is still used in your work. I would like to ask if you have tried not using the softmax method, but can you still achieve vision tasks without losing too much accuracy?

I have tried the cosformer(cos + relu), which lowered the visual task by 10 points.

By the way, I provide the corresponding code when the q and k dimensions are different.
sink_incoming = 1.0 / torch.sum(
    (query + self.eps) * (key.sum(dim=2, keepdim=True).repeat(1, 1, L, 1) + self.eps), dim=-1)
source_outgoing = 1.0 / torch.sum(
    (key + self.eps) * (query.sum(dim=2, keepdim=True).repeat(1, 1, S, 1) + self.eps), dim=-1)
conserved_sink = torch.sum((query + self.eps) * (
        (key * source_outgoing[:, :, :, None]).sum(dim=2, keepdim=True).repeat(1, 1, L, 1) + self.eps), dim=-1)
conserved_source = torch.sum((key + self.eps) * (
        (query * sink_incoming[:, :, :, None]).sum(dim=2, keepdim=True).repeat(1, 1, S, 1) + self.eps), dim=-1)

**Comment by wuhaixu2016** [comment]
Hi, thanks for your insightful question.

As we stated in our other paper: Flowformer (https://github.com/thuml/Flowformer), softmax is essential to avoid trivial attention, which is why still keep this operation in our design. However, in our paper, the usage will only take linear complexity, which is much faster than canonical attention.

We also compare with cosformer in Flowformer paper, you will see that the local inductive bias of cosformer will lead to a very weird attention map.
